auth/src/main/java/com/firebase/ui/auth/ui/email/RecoverPasswordProgressState.java

```diff
@@ -0,0 +1,67 @@
+package com.firebase.ui.auth.ui.email;
+
+import android.support.annotation.NonNull;
+import android.support.annotation.RestrictTo;
+
+import com.firebase.ui.auth.data.model.ProgressState;
+
+/**
+ * Password recovery state.
+ * <p>
+ * When this state is successful, the email who's password was reset can be retrieved.
+ */
+@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+public final class RecoverPasswordProgressState extends ProgressState {
```

Comment:
Ok you're probably going to kill me for suggesting another direction here ... but I am learning a lot with each iteration!

I saw someone the other day define a class basically like this (simplified code):

```java
public class Resource<T> {

   State state; // some enum
   T value;
   Exception e;

}
```

So then recover password screen would watch a single live data that is `Resource<String>` where the `String` is the email.  And the `Observer` would get a series of `Resource` objects with states like `STATE_NONE`, `STATE_LOADING`, `STATE_SUCCESS/STATE_FAILURE`.  So it's kind of like `Task` but adapted for the world of `LiveData`.

In this way you don't need a subclass of `ProgressState` for each screen that loads something just a different `<T>`, which could be a compound object you define if you're waiting for a few results.

What do you think?  Cool?  Stupid?

Reply:
Ooooo, no, I really like that idea! 😁 It'll even make sure that we write proper model objects if we need to pass around a multi-arg result. (Let's agree not to use `Pair` 😉) Also, did you learn this from a talk or something I could look at too? Sounds really cool!